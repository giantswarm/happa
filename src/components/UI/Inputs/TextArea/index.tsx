import { FormField, FormFieldProps, TextArea as Input } from 'grommet';
import * as React from 'react';

interface ITextAreaProps extends React.ComponentPropsWithoutRef<typeof Input> {
  /**
   * The description text displayed above the input.
   */
  label?: FormFieldProps['label'];
  /**
   * Props to be passed to the text input wrapper.
   */
  contentProps?: FormFieldProps['contentProps'];
  /**
   * Props to be passed to the form field.
   */
  formFieldProps?: FormFieldProps & { className?: string };
  /**
   * Whether the input is required or not.
   */
  required?: Exclude<FormFieldProps['required'], { indicator: boolean }>;
  /**
   * An error to be displayed below the input.
   */
  error?: FormFieldProps['error'];
  /**
   * A description to be displayed below the input.
   */
  info?: FormFieldProps['info'];
  /**
   * A help text to be displayed between the label
   * and the input.
   */
  help?: FormFieldProps['help'];
  /**
   * The margin between the form field and other elements.
   */
  margin?: FormFieldProps['margin'];
  /**
   * Whether the input should have some extra padding or not.
   */
  pad?: FormFieldProps['pad'];
}

const TextArea = React.forwardRef<HTMLTextAreaElement, ITextAreaProps>(
  (
    {
      id,
      label,
      contentProps,
      formFieldProps,
      disabled,
      required,
      error,
      info,
      help,
      name,
      margin,
      pad,
      ...props
    },
    ref
  ) => {
    return (
      <FormField
        htmlFor={id}
        label={label}
        contentProps={contentProps}
        disabled={disabled}
        required={required}
        name={name}
        error={error}
        info={info}
        help={help}
        margin={margin}
        pad={pad}
        {...formFieldProps}
      >
        <Input
          {...props}
          id={id}
          ref={ref}
          disabled={disabled}
          required={required}
          name={name}
        />
      </FormField>
    );
  }
);

TextArea.defaultProps = {
  size: 'medium',
  rows: 5,
  resize: 'vertical',
};

export default TextArea;
