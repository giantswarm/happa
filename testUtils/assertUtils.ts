import { MatcherFunction, Nullish } from '@testing-library/react';

type Query = (matcher: MatcherFunction) => HTMLElement;

/**
 * Decorate an existing query to match text across multiple
 * elements.
 * @param query
 */
export function withMarkup(query: Query) {
  return (text: string) => {
    return query((_, node: Nullish<Element>) => {
      if (!node) return false;

      const hasText = (n: Element) => n.textContent === text;
      const hasTextInChildren = Array.from(node.children).some((child) =>
        hasText(child)
      );

      return hasText(node) && !hasTextInChildren;
    });
  };
}
