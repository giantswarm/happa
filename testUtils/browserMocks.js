// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('isomorphic-fetch');

const localStorageMock = (function () {
  let store = {};

  return {
    getItem: function (key) {
      return store[key] || null;
    },
    setItem: function (key, value) {
      store[key] = value;
    },
    removeItem: function (key) {
      delete store[key];
    },
    clear: function () {
      store = {};
    },

    // replaceWith is not part of the localStorage spec, but useful for us
    // in testing to set the full localStorage to a known set of values.
    replaceWith: function (newStore) {
      store = newStore;
    },
  };
})();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

// Adjust the console logging behaviour during test runs.
/* eslint-disable no-console */
global.console = {
  log: console.log,
  error: console.error,
  warn: jest.fn(), // warnings are surpressed.
  info: console.info,
  debug: console.debug,
};
/* eslint-enable no-console */

global.fetch = fetch;
