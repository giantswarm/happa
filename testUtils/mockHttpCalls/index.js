/***** Constants *****/
export * from './constantsAndHelpers';

/***** Responses *****/

// Info
export * from './info';

// User
export * from './user';

// Auth
export * from './auth';

// Clusters
export * from './clusters';

// Organizations
export * from './organizations';

// Status
export * from './status';

// Apps
export * from './apps';

// Node Pools
export * from './nodePools';

// Releases
export * from './releases';

// App Catalogs
export * from './appCatalogs';
