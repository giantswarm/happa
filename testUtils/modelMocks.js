import * as authSelectors from 'selectors/authSelectors';

jest.mock('model/services/giantSwarm');
jest.mock('selectors/authSelectors');

// eslint-disable-next-line no-import-assign
authSelectors.selectAuthToken = jest.fn(
  (_, state) => state.app.loggedInUser.token
);
