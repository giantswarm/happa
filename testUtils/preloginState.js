import { Providers } from 'shared/constants';
import { MainRoutes } from 'shared/constants/routes';

// This is what the state looks like when someone brand new arrives at the site
// and they are not logged in at all.

export default {
  router: {
    location: {
      pathname: MainRoutes.Login,
      search: '',
      hash: '',
      key: '0fgz24',
    },
    action: 'POP',
  },
  main: {
    selectedOrganization: null,
    selectedClusterID: null,
    firstLoadComplete: false,
    loggedInUser: null,
    permissions: {},
    info: {
      general: {
        installation_name: '',
        availability_zones: null,
        provider: Providers.AWS,
      },
      stats: {
        cluster_creation_duration: {
          median: 0,
          p25: 0,
          p75: 0,
        },
      },
      workers: {
        count_per_cluster: {
          max: null,
          default: 0,
        },
      },
    },
  },
  errors: {},
  metadata: {
    version: {
      current: '',
      new: null,
      lastCheck: 0,
      timer: 0,
    },
  },
  entities: {
    catalogs: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
    },
    clusters: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      v5Clusters: [],
      idsAwaitingUpgrade: {},
    },
    organizations: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      credentials: {
        lastUpdated: 0,
        isFetching: false,
        items: [],
        showForm: false,
      },
    },
    releases: {
      error: null,
      isFetching: false,
      items: {},
    },
    users: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      invitations: {
        lastUpdated: 0,
        isFetching: false,
        items: {},
      },
    },
    nodePools: {
      items: {},
      isFetching: false,
    },
    clusterLabels: {
      requestInProgress: false,
      error: null,
    },
  },
  modal: {
    visible: false,
    templateValues: {},
    template: '',
  },
};
