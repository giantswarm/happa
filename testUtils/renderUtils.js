import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
