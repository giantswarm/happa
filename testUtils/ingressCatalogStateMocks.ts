import { Constants } from 'shared';
import { appCatalogsResponse } from 'testUtils/mockHttpCalls';

const ingressCatalog = appCatalogsResponse.find(
  (c) => c.metadata.name === Constants.INSTALL_INGRESS_TAB_APP_CATALOG_NAME
);
export const catalogsState = {
  entities: {
    catalogs: {
      items: {
        [Constants.INSTALL_INGRESS_TAB_APP_CATALOG_NAME]: ingressCatalog,
      },
    },
  },
};
