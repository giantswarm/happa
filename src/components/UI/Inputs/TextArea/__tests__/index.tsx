import { renderWithTheme } from 'testUtils/renderUtils';

import TextArea from '..';

describe('TextArea', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(TextArea, {
      value: 'The quick brown fox jumps over the lazy dog',
      placeholder: 'Please write a story',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a label', () => {
    const { container } = renderWithTheme(TextArea, {
      value: 'The quick brown fox jumps over the lazy dog',
      label: 'A helpful area of text',
      id: 'some-text',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a validation error', () => {
    const { container } = renderWithTheme(TextArea, {
      value: 'The quick brown fox jumps over the lazy dog',
      error: 'There is something very wrong!',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a help message', () => {
    const { container } = renderWithTheme(TextArea, {
      value: 'The quick brown fox jumps over the lazy dog',
      help: 'A helpful message',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with an info message', () => {
    const { container } = renderWithTheme(TextArea, {
      value: 'The quick brown fox jumps over the lazy dog',
      info: 'Some useful information',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled input', () => {
    const { container } = renderWithTheme(TextArea, {
      value: 'The quick brown fox jumps over the lazy dog',
      disabled: true,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
