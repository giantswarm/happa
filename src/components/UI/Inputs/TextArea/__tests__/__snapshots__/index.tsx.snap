// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders a disabled input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hRVLan"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 eKUXDU"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dWhfdE FormField__FormFieldBox-sc-m9hood-0 iLyaEy"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 gGtqur FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <textarea
          class="StyledTextArea-sc-17i3mwp-0 iOCFJY TextArea__StyledInput-sc-ixw2m3-0 eHLMtF"
          disabled=""
          rows="5"
        >
          The quick brown fox jumps over the lazy dog
        </textarea>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextArea renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hRVLan"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 eKUXDU"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dWhfdE FormField__FormFieldBox-sc-m9hood-0 iLyaEy"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 lmnuvp FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <textarea
          class="StyledTextArea-sc-17i3mwp-0 jUpZct TextArea__StyledInput-sc-ixw2m3-0 eHLMtF"
          placeholder="Please write a story"
          rows="5"
        >
          The quick brown fox jumps over the lazy dog
        </textarea>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextArea renders an input with a help message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hRVLan"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 eKUXDU"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dWhfdE FormField__FormFieldBox-sc-m9hood-0 iLyaEy"
    >
      <span
        class="StyledText-sc-1sadyjn-0 dkbpSN"
      >
        A helpful message
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 lmnuvp FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <textarea
          class="StyledTextArea-sc-17i3mwp-0 jUpZct TextArea__StyledInput-sc-ixw2m3-0 eHLMtF"
          rows="5"
        >
          The quick brown fox jumps over the lazy dog
        </textarea>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextArea renders an input with a label 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hRVLan"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 eKUXDU"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dWhfdE FormField__FormFieldBox-sc-m9hood-0 iLyaEy"
    >
      <label
        class="StyledText-sc-1sadyjn-0 julVzF"
        for="some-text"
      >
        A helpful area of text
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 lmnuvp FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <textarea
          class="StyledTextArea-sc-17i3mwp-0 jUpZct TextArea__StyledInput-sc-ixw2m3-0 eHLMtF"
          id="some-text"
          rows="5"
        >
          The quick brown fox jumps over the lazy dog
        </textarea>
      </div>
    </div>
  </div>
</div>
`;

exports[`TextArea renders an input with a validation error 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hRVLan"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 eKUXDU"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dWhfdE FormField__FormFieldBox-sc-m9hood-0 iLyaEy"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 hDnXgl FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <textarea
          class="StyledTextArea-sc-17i3mwp-0 jUpZct TextArea__StyledInput-sc-ixw2m3-0 eHLMtF"
          rows="5"
        >
          The quick brown fox jumps over the lazy dog
        </textarea>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 YswyP"
      >
        There is something very wrong!
      </span>
    </div>
  </div>
</div>
`;

exports[`TextArea renders an input with an info message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hRVLan"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 eKUXDU"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dWhfdE FormField__FormFieldBox-sc-m9hood-0 iLyaEy"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 lmnuvp FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <textarea
          class="StyledTextArea-sc-17i3mwp-0 jUpZct TextArea__StyledInput-sc-ixw2m3-0 eHLMtF"
          rows="5"
        >
          The quick brown fox jumps over the lazy dog
        </textarea>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 hleChS"
      >
        Some useful information
      </span>
    </div>
  </div>
</div>
`;
